src/content/marina/marinaBroker.ts

```diff
@@ -380,28 +397,42 @@ export default class MarinaBroker extends Broker<keyof Marina> {
           const network = await this.appRepository.getNetwork();
           if (!network) throw new Error('Network not set up');
           const coins = await this.walletRepository.getUtxos(network, ...accountsIDs);
-          const results: Utxo[] = [];
+          const utxos: Utxo[] = [];
+          const unblinder = new WalletRepositoryUnblinder(
+            this.walletRepository,
+            this.appRepository,
+            this.assetRepository,
+            await zkpLib()
+          );
           for (const coin of coins) {
-            const witnessUtxo = await this.walletRepository.getWitnessUtxo(coin.txID, coin.vout);
-            let scriptDetails: ScriptDetails | undefined;
-
-            if (witnessUtxo) {
-              const script = witnessUtxo.script.toString('hex');
-              const { [script]: details } = await this.walletRepository.getScriptDetails(
-                witnessUtxo?.script.toString('hex')
-              );
-              scriptDetails = details;
+            const utxo = await this.unblindedOutputToUtxo(coin);
+            // check if the coins need unblinding
+            if (
+              coin.blindingData ||
+              !utxo.witnessUtxo ||
+              !utxo.witnessUtxo.rangeProof ||
+              !utxo.witnessUtxo.surjectionProof ||
+              !(utxo.witnessUtxo.rangeProof.length > 0) ||
+              !(utxo.witnessUtxo.surjectionProof.length > 0)
+            ) {
+              utxos.push(utxo);
+              continue;
             }
 
-            results.push({
-              ...coin,
-              txid: coin.txID,
-              scriptDetails,
-              witnessUtxo,
-            });
+            // if need unblinding, unblind, cache result, and add to the list
+            const unblinded = (await unblinder.unblind(utxo.witnessUtxo)).at(0);
+            if (unblinded instanceof Error || !unblinded) {
+              console.warn('unblinding failed', unblinded);
```

Comment:
Shall we schedule a retry instead of only logging?

Reply:
or maybe in this case throwing an error back to the requester

Reply:
I made it on purpose because the `Utxo` type also have a `blindingPrivateKey` field & `blindingData` are optional thus it's also possible to generate them or use the blindingPrivateKey to unblind the input while spending the utxo

Reply:
So ok you saying its fine to return the wrong data (ie missing the blindingData), as the user can retry to get more recent coins?

Reply:
I mean most of the time getCoins() is used with coin select (cause we already have getBalances, getTxs for other types of data) so I think it's ok to return the utxo even if marina was not able to unblind it using masterBlindingKey (we don't know, maybe the app is using a special blinding process)